pytorch_lightning/loops/batch/training_batch_loop.py

```diff
@@ -137,6 +137,10 @@ def advance(self, batch, batch_idx):
             if result:
                 self.batch_outputs[0].append(deepcopy(result.result_collection))
 
+                # clear reference to this step's training loss so that it can be garbage
+                # collected before the next training step
+                trainer._results.minimize = None
```

Comment:
Is this only a problem with manual optimization?

This should probably be moved to `LoggerConnector.on_train_split_start`.

cc @tchaton this is related to avoiding the `ResultCollection` as a vehicle to pass data between loops and iterations as the state is harder to manage (as we saw with deepcopy)

Reply:
No, it happens with automatic optimization. My use case is a bit particular. I am training neural ODEs that can have varying memory requirements per batch and sometimes the memory usage spikes causing a CUDA OOM and aborting the training. To alleviate that a bit and filter our problematic batches, my module looks basically like the following
```python
import pytorch_lightning as pl
from pytorch_lightning.utilities.memory import (
    garbage_collection_cuda,
    is_cuda_out_of_memory,
)

class ExampleModule(pl.LightningModule):
    def training_step(self, *args):
        try:
            return { "loss": ... }
        except RuntimeError as e:
            if is_cuda_out_of_memory(e):
                return None
            else:
                raise

    def backward(self, *args, **kwargs):
        try:
            return super().backward(*args, **kwargs)
        except RuntimeError as e:
            if is_cuda_out_of_memory(e):
                return None
            else:
                raise
```
What can happen then is that a CUDA OOM occurs during the backward pass. In that case the context tensors of the backwards graph do not get freed. Now, if I have a loss with a huge backwards graph (let's say 99% of CUDA memory), `backward` will surely fail, leaving the whole graph in memory. That graph is still attached to the loss that is stored in `trainer._results.minimize`. Because that reference keeps the graph and memory alive throughout the next training step leaving only 1% of CUDA memory, no forward pass will ever succeed again and `trainer._results.minimize` will never be overwritten. This PR fixes that by clearing that reference as soon as possible.

All in all, I agree that the root cause is this mutable `trainer._results` object that is modified and copied for some reason instead of just creating new instances.

Reply:
Yes, we need to refactor the code, but we can merge this one as a hotfix I guess.

Reply:
Still as, I said:

> This should probably be moved to LoggerConnector.on_train_split_start.

So it gets freed before